rfd/0159-webasset-cache.md

```diff
@@ -0,0 +1,165 @@
+but should be made explicit at least in the documentation for this feature. 
+
+For example, if auth is 15.0.1 and then we upgrade to 15.1.1, proxy cannot skip
+and go straight to 15.1.2 (or behind at 15.1.0) as it will then not have its version
+of the webassets uploaded by auth.
+
+#### File cleanup
+Our current webasset bundle is ~7mb. The first cause for concern is how to
+mitigate an s3 bucket filling up indefinitely with new versions of the
+webassets. Due to the chaotic nature of when/how often different users choose to
+upgrade, I believe it it out of scope for Teleport to have any control of bucket
+cleanup. However, we can provide some general suggestions and documentation that
+we think would work best. One of the things Teleport _can_ provide is a
+`TeleportVersion` tag for each webasset file (assuming the compatible storage
+supports tags). This will give the users some insight into which version a
+specific file is supporting.
+
+The suggested way to cleanup files is to set a [lifecycle rule](https://docs.aws.amazon.com/AmazonS3/latest/userguide/object-lifecycle-mgmt.html) based on expiration. This will
+let files naturally cleanup themselves after a specified time. If a file is deleted due to this rule for an auth version still
+in use, then the auth will just reupload the file during the next heartbeat.
+
+According to AWS, the lifecycle rules are only checked [once per day](https://repost.aws/knowledge-center/s3-lifecycle-rule-delay), and are rounded to midnight UTC of the next day when an object would be "expired".  However, different storages may have different timings here, so it's best the heartbeat runs frequently. (every 10 minutes?)
+
+Ultimately, the user can setup their cleanup however they wish, and Teleport is responsible for filling up the bucket.
+
+
+
+#### Serving the missing webassets
+Currently, our static file serving looks [like this](https://github.com/gravitational/teleport/blob/master/lib/web/apiserver.go#L513C4-L518)
+```go
+if strings.HasPrefix(r.URL.Path, "/web/app") {
+    fs := http.FileServer(cfg.StaticFS)
+
+    fs = makeGzipHandler(fs)
+    fs = makeCacheHandler(fs)
+
+    http.StripPrefix("/web", fs).ServeHTTP(w, r)
+}
+```
+
+nothing too complicated besides setting some headers, compression, and then
+serving whatever file is requested. The change would be instead of serving,
+first check if the file exists in the embedded file system. If it does, send as
+usual. If it doesn't, ask auth for the missing webasset by name, auth downloads
+from the bucket, and then ships back to proxy to serve.
+```go
+if strings.HasPrefix(r.URL.Path, "/web/app") {
+	// do everything the same if the file exists
+	if (fileExists) {
+		fs := http.FileServer(cfg.StaticFS)
+
+		fs = makeGzipHandler(fs)
+		fs = makeCacheHandler(fs)
+
+		http.StripPrefix("/web", fs).ServeHTTP(w, r)
+        return
+	}
+	writer := &MemoryWriterAt{}
+	cfg.WebassetHandler.Download(cfg.Context, fileName, writer)
```

Comment:
Yes, you would receive the same error/404 if a file isn't found in the bucket after checking. If that is the case, it would be a great spot to show the "Refresh" popup but that hasn't been approved by design yet.